Dockerfile

```diff
@@ -1,53 +1,78 @@
-FROM ubuntu:16.04
-RUN apt-get update && apt-get install -y \
-    git \
+FROM ubuntu:18.04
+
+ENV GO_VERSION=1.10.2
+
+# Install JDK via PPA: https://github.com/franzwong/til/blob/master/java/silent-install-oracle-jdk8-ubuntu.md
+RUN apt-get update && \
+    apt-get install -y python3-software-properties software-properties-common && \
+    add-apt-repository -y ppa:webupd8team/java && \
+    apt-get update && \
+    echo "oracle-java8-installer shared/accepted-oracle-license-v1-1 select true" | debconf-set-selections && \
+    apt-get install -y \
     curl \
+    git \
     jq \
-    default-jdk \
+    libssl-dev \
+    oracle-java8-installer \
     python-pip \
     python3 \
     python3-dev \
     python3-pip \
     rsync \
     tox \
     software-properties-common \
-    python-software-properties \
-    libssl-dev \
     upx-ucl \
     wget \
     zip && \
-    rm -rf /var/lib/apt/lists/*
+    rm -rf /var/lib/apt/lists/* && \
+    java -version && \
+    curl -O https://storage.googleapis.com/golang/go${GO_VERSION}.linux-amd64.tar.gz && \
+    tar -xf go${GO_VERSION}.linux-amd64.tar.gz && \
+    mv go /usr/local && \
+    rm -f go${GO_VERSION}.linux-amd64.tar.gz
 
-# Install go 1.8.5
-RUN curl -O https://storage.googleapis.com/golang/go1.8.5.linux-amd64.tar.gz && \
-    tar -xf go1.8.5.linux-amd64.tar.gz && \
-    mv go /usr/local
 ENV PATH=$PATH:/usr/local/go/bin
 RUN go version
 
 # AWS CLI for uploading build artifacts
 RUN pip3 install awscli
-# Install the testing dependencies
+# Install the lint+testing dependencies
 COPY test_requirements.txt test_requirements.txt
-RUN pip3 install -r test_requirements.txt
-# shakedown and dcos-cli require this to output cleanly
+RUN pip3 install --upgrade -r test_requirements.txt
+
+# Get DC/OS CLI
+RUN curl -O https://downloads.dcos.io/binaries/cli/linux/x86-64/dcos-1.12/dcos && \
+    chmod +x dcos && \
+    mv dcos /usr/local/bin && \
+    dcos --version
+
+# dcos-cli and lint tooling require this to output cleanly
 ENV LC_ALL=C.UTF-8 LANG=C.UTF-8
 # use an arbitrary path for temporary build artifacts
 ENV GOPATH=/go-tmp
 # make a dir for holding the SSH key in tests
 RUN mkdir /root/.ssh
 
+# Copy all of the repo into the image, then run some build/lint commands against the copy to heat up caches. Then delete the copy.
+RUN mkdir /tmp/repo/
+COPY / /tmp/repo/
+# gradlew: Heat up jar cache. pre-commit: Heat up lint tooling cache.
+RUN cd /tmp/repo/ && \
+    ./gradlew testClasses && \
+    pre-commit install-hooks && \
```

Comment:
let's add the `git init && \` line here as with #2645 as it should unblock this.